File: 字符串/6. 实现strStr() - 28/Find the Index of the First Occurrence in a String.py

# 23/03/29
# https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/

# 讲解视频：https://www.bilibili.com/video/BV1AY4y157yL/?spm_id_from=333.337.search-card.all.click&vd_source=7e6e2d618beb86d965f4d817126fc72f

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:

        def getNextList(pattern: list()) -> list:

            nxt = [0]
            lengthOfPrefix = 0
            pointer = 1

            while pointer < len(pattern):
                if pattern[pointer] == pattern[lengthOfPrefix]:
                    lengthOfPrefix += 1
                    nxt.append(lengthOfPrefix)
                    pointer += 1
                elif lengthOfPrefix == 0:
                    nxt.append(0)
                    pointer += 1
                else:
                    lengthOfPrefix = nxt[lengthOfPrefix - 1]
            return nxt
        
        nxt = getNextList(needle)
        motherPointer, childPointer = 0, 0
        res = -1

        while motherPointer < len(haystack):
            if haystack[motherPointer] == needle[childPointer]:
                motherPointer += 1
                childPointer += 1
            elif childPointer > 0:
                childPointer = nxt[childPointer - 1]
            else:
                motherPointer += 1
            
            if childPointer == len(needle):
                res = motherPointer - childPointer
                break
        
        return res

File: 字符串/6. 实现strStr() - 28/Find the Index of the First Occurrence in a String.cpp
/*
    23/03/29
    https://leetcode.cn/problems/find-the-index-of-the-first-occurrence-in-a-string/
 */

#include <iostream>
#include <string>

using namespace std;

/*
    solution: KMP
    result: ✅
    n: length of pattern
    m: length of string
    TC: O(n + m)
    SC: O(n)
 */

class Solution {
public:

    void buildNext(int* next, const string& s) {
        int lengthOfPrefix = 0;
        int pointer = 1;
        next[0] = 0;

        while (pointer < s.size()) {
            if (s[pointer] == s[lengthOfPrefix]) {
                lengthOfPrefix += 1;
                next[pointer++] = lengthOfPrefix;
            } else if (lengthOfPrefix == 0) {
                next[pointer++] = 0;
            } else {
                lengthOfPrefix = next[lengthOfPrefix - 1];
            }
        }
    }

    int strStr(string haystack, string needle) {
        if (needle.size() == 0) {return 0;}

        int next[needle.size()];
        buildNext(next, needle);
        int motherPointer = 0;
        int childPointer = 0;
        while (motherPointer < haystack.size()) {
            if (haystack[motherPointer] == needle[childPointer]) {
                motherPointer += 1;
                childPointer += 1;
            } else if (childPointer > 0){
                childPointer = next[childPointer - 1];
            } else {
                motherPointer += 1;
            }

            if (childPointer == needle.size()) {
                return motherPointer - childPointer;
            }
        }
        return -1;
    }
};